### PR TITLE
Prevent grad scale from overflowing

### LINKED ISSUE #98828: torch.cuda.amp.GradScaler._scale can overflow
### 🐛 Describe the bug

For networks where the loss is small, it can happen that the gradscaler overflows before the gradients become infinite. 
```python
import torch
import torch.nn as nn

net = nn.Linear(5,1).cuda()
optimizer = torch.optim.Adam(net.parameters())
scaler = torch.cuda.amp.GradScaler(growth_interval=1)
for t in range(1000):
    optimizer.zero_grad()
    x = torch.randn(1,5).cuda()
    y = 0.0001 * torch.randn(1,1).cuda()
    l = ((net(x)-y)**2).mean()
    scaler.scale(l).backward()
    scaler.step(optimizer)
    scaler.update()
    print(scaler._scale)
```

Limiting the growth of the gradscaler, or using float64 instead of float32 seems like a reasonable idea.

### Versions

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==2.0.0
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### PR DESCRIPTION
**Pull request by janeyx99**
Fixes #98828 by capping the growth in the kernel

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98876

cc @ngimel

### CODE REVIEW

**Comment by janeyx99** [comment]
cc @crcrpar to make sure this is an okay approach.

**Review by ngimel — APPROVED** [review]

File: test/test_cuda.py [hunk 1]
```diff
@@ -2521,6 +2521,20 @@ def _grad_scaling_autocast_fused_optimizers(self, optimizer_ctor, optimizer_kwar
...
+        scaler = torch.cuda.amp.GradScaler(growth_interval=1)
+        for _ in range(150):
+            optimizer.zero_grad()
+            x = torch.randn(1, 5).cuda()
+            y = 1e-30 * torch.randn(1, 1).cuda()
+            l = ((model(x) - y)**2).mean()
+            scaler.scale(l).backward()
+            scaler.step(optimizer)
+            scaler.update()
+        assert(scaler._scale is not float('inf') and scaler._scale is not float('nan'))
```
Comment (crcrpar) [inline test/test_cuda.py:2536]: Hmm, 150 iterations sound lot while the computation is quite light.
What do you think about initializing `GradScaler` with a ridiculously large `growth_factor` and `growth_interval=1` so that only one `scaler.step` would cause overflow?

ref: https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler

Comment (janeyx99) [reply test/test_cuda.py]: I've contrived something similar 😛
